File: abc418/e/main.rb
#!/usr/bin/env ruby

# acc n ContestID
# oj t -c " ruby main.rb" -d test
# acc s 


# require "ac-library-rb/priority_queue"
# require "ac-library-rb/segtree"
# require "ac-library-rb/dsu"

# pq = AcLibraryRb::PriorityQueue.new

# require 'prime'
# pd = Prime.prime_division(a)

$debug = !ARGV[0].nil?

N = STDIN.gets.chomp.to_i
XYS = (1..N).map do
  STDIN.gets.chomp.split.map(&:to_i)
end


ds = {}
ds_minus = {}
XYS.sort.combination(2) do |(x1, y1), (x2, y2)|
  # pp(x1:, y1:, x2:, y2:) if $debug
  dx = x2 - x1
  dy = y2 - y1
  g = dx.gcd(dy)
  dx /= g
  dy /= g
  ds[dx] ||= Hash.new(0)
  ds[dx][dy] += 1

  ds_minus[x2 + x1] ||= Hash.new(0)
  ds_minus[x2 + x1][y2 + y1] += 1
end
pp(ds: ds.sort) if $debug

r = 0
r_minus = 0
ds.each do |dx, dys|
  dys.each do |dy, c|
    r += c * (c - 1) / 2
  end
end
ds_minus.each do |dx, dys|
  dys.each do |dy, c|
    r_minus += c * (c - 1) / 2
  end
end

pp(r:, r_minus:) if $debug
puts r - r_minus

File: abc418/e/main.py
#!/usr/bin/env python3
import sys
from math import gcd
from itertools import combinations
from pprint import pprint

# 入力読み取り
N = int(sys.stdin.readline().strip())
XYS = [tuple(map(int, sys.stdin.readline().split())) for _ in range(N)]

# 方向ベクトルを分類
ds = {}
for (x1, y1), (x2, y2) in combinations(sorted(XYS), 2):
    dx = x2 - x1
    dy = y2 - y1
    g = gcd(dx, dy)
    dx //= g
    dy //= g

    # Ruby の ds[dx][dy][g] に相当する辞書構造
    ds.setdefault(dx, {}).setdefault(dy, {}).setdefault(g, 0)
    ds[dx][dy][g] += 1

# 計算部分
r = 0
r_minus = 0
for dx, dys in ds.items():
    for dy, gs in dys.items():
        c_sum = 0
        for g, c in gs.items():
            c_sum += c
            r_minus += c * (c - 1) // 2
        r += c_sum * (c_sum - 1) // 2

# 出力
print(r - r_minus // 2)
